clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -407,6 +408,7 @@ public class KafkaAdminClient extends AdminClient {
     private final ExponentialBackoff retryBackoff;
     private final boolean clientTelemetryEnabled;
     private final MetadataRecoveryStrategy metadataRecoveryStrategy;
+    private final Map<TopicPartition, Integer> partitionLeaderCache;
```

Comment:
Do we need to manage the memory usage of this cache if it's being used to optimize long-running admin operations? Maybe we could set a size limit, or clean it up when the metadata expires (reusing the existing `metadata.max.age.ms` configuration)?

Reply:
I've been thinking about this and I'm not so sure.

With `metadata.max.age.ms`, there is typically a reason to need to refresh the metadata, such as discovering new topics that match a wildcard or learning about new partitions on an existing topic. Here's a cache entry is good until the partition leader changes.

With this cache, it could contain an entry for each topic-partition whose leader information was required in order to perform admin actions using the PartitionLeaderStrategy. If an entry is stale, the fulfilment stage will fail with `NOT_LEADER_OR_FOLLOWER` and the lookup will be performed again. Would it really be better to mistrust the cached data after say 5 minutes and lookup again anyway? I'm not convinced that's true.

Reply:
> With this cache, it could contain an entry for each topic-partition whose leader information was required in order to perform admin actions using the PartitionLeaderStrategy. If an entry is stale, the fulfilment stage will fail with NOT_LEADER_OR_FOLLOWER and the lookup will be performed again. Would it really be better to mistrust the cached data after say 5 minutes and lookup again anyway? I'm not convinced that's true.

My another concern was that the cluster might experience a lot of topic deletions, leading to a cache filled with garbage that we can't collect. However, this might be fine, as those nonexistent partitions don't take up much memory, and the time complexity of `HashMap` operations is `O(1)`.